core/src/ppu.cpp

```diff
@@ -36,6 +41,16 @@ uint8_t Ppu::read_byte(uint16_t addr) {
         clear_vblank_flag();
     } else if (addr == kOamData) {
         byte = oam_data_[registers_->oamaddr];
+    } else if (addr == kPpuData) {
+        byte = mmu_->read_byte(registers_->vram_addr);
+        if (registers_->vram_addr < kFirstPaletteData) {
+            const uint8_t tmp_buffer = read_buffer_;
+            read_buffer_ = byte;
+            byte = tmp_buffer;
+        } else {
+            // TODO(ML): Update read_buffer_
```

Comment:
A bit hard to interpret what should happen here. In the documentation it says:

> Reading palette data from $3F00-$3FFF works differently. The palette data is placed immediately on the data bus, and hence no dummy read is required. Reading the palettes still updates the internal buffer though, but the data placed in it is the mirrored nametable data that would appear "underneath" the palette. (Checking the PPU memory map should make this clearer.)

Any suggestions?

Reply:
Perhaps just:
read_buffer = byte;, and return byte?

Regarding the mirroring, then this post (https://forums.nesdev.com/viewtopic.php?f=3&t=18627) suggests reading from vram_addr - 0x1000 instead in that case.